### PR TITLE
cache columns row to avoid malloc/free penalty

### PR DESCRIPTION
**Pull request by adizaimi**
* Avoiding call to newsql_alloc_row() gets 1.5% savings from not mallocing and deallocing row. 
* Not mallocing but using alloca in make_retrow() gives 0.5% savings.

### CODE REVIEW

File: db/sqlinterfaces.c [hunk 1]
```diff
@@ -1185,11 +1185,36 @@ int newsql_send_last_row(struct sqlclntstate *clnt, int is_begin,
                          malloc, 1);
 }
 
+static int newsql_realloc_row(int ncols, CDB2SQLRESPONSE__Column ***columns, int *columns_count)
+{
+    if(ncols <= *columns_count)
+        return 0; /* already have more columns set up; nothing to do */
+
+printf("AZ: realloc from %d to %d\n", *columns_count, ncols);
```
Comment (ldennis) [inline db/sqlinterfaces.c]: remove trace

File: db/sqlinterfaces.c [hunk 2]
```diff
@@ -1185,11 +1185,36 @@ int newsql_send_last_row(struct sqlclntstate *clnt, int is_begin,
...
+    int newtotsize = arrsz + ncols * sizeof(CDB2SQLRESPONSE__Column);
+
+    CDB2SQLRESPONSE__Column **loc_cols = *columns;
+    char *ptr = realloc(loc_cols, newtotsize);
+    if (!ptr) {
+        return -1;
+    }
+
+    loc_cols = (CDB2SQLRESPONSE__Column **) ptr;
+    for (int i = 0; i < ncols; i++) {
```
Comment (ldennis) [inline db/sqlinterfaces.c]: The loop can start from `*columns_count` instead?

Comment (adizaimi) [reply db/sqlinterfaces.c]: We realloc, so unfortunately we have to reassign new pointer locations for the individual colums.

Comment (ldennis) [reply db/sqlinterfaces.c]: Sorry, you were right.

**Comment by adizaimi** [comment]
thanks Lingzhi.

**Pull request merged**